torch_xla_py/parallel_loader.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import division
+from __future__ import print_function
+
+import collections
+from six import itervalues
+import multiprocessing.dummy
+import os
+import threading
+import time
+import torch
+import torch.nn as nn
+import torch_xla
+import torch_xla_py.utils as xu
+import torch_xla_py.keyd_queue as kq
+
+
+class ParallelLoader(object):
+
+  def __init__(self, loader, batch_size, devices, prefetch_size=4, batchdim=0):
+    self._loader = loader
+    self._prefetch_size = prefetch_size
+    self._batch_size = batch_size
+    self._devices = list(devices)
+    self._device_indices = dict()
+    for i, device in enumerate(self._devices):
+      self._device_indices[device] = i
+    self._batch_number = 0
+    self._batchdim = batchdim
+    self._done = False
+    self._lock = threading.Lock()
+    self._loader_queue = kq.Queue(maxsize=self._prefetch_size)
+    self._queue = kq.KeydQueue(maxsize=self._prefetch_size)
+    self._worker_count = 0
+    self._data = None
+    self._device_slices = None
+    thread = threading.Thread(target=self._loader_worker)
+    thread.daemon = True
+    thread.start()
+    for _ in range(0, prefetch_size):
+      thread = threading.Thread(target=self._worker)
+      thread.daemon = True
+      thread.start()
+
+  def __iter__(self):
+    return self
+
+  def __next__(self):
+    return self.next()
+
+  def next(self):
+    item = self._queue.get(self._batch_number)
+    if item is None:
+      raise StopIteration
+    self._data, target, self._device_slices = item
+    self._batch_number += 1
+    return self._batch_number - 1, (self._data, target)
+
+  def close(self):
+    self._done = True
+    self._queue.close()
+    self._loader_queue.close()
+
+  def to(self, data, device):
+    assert data is self._data
+    return self._device_slices[self._device_indices[device]]
+
+  def _up_workers(self, count):
+    with self._lock:
+      self._worker_count += count
+      return self._worker_count
+
+  def _loader_worker(self):
+    batch_number = 0
+    for (data, target) in self._loader:
+      if data.size()[self._batchdim] != self._batch_size or self._done:
```

Comment:
Is this meant to be used with lazy tensors only? Because we don't have to skip batches of "wrong" size in that case, we can handle it.

Reply:
Yes, it is.
No, we cannot handle it, as CRSs wants same shapes ;)

Reply:
Ah, too bad. Is it relatively easy to lift the CRS constraint? Not something we need to address now, but it looks feasible to do so and not have the last batch problem at all.